File: benchmarks/v3/scenarios/05-java-saga/with-mcp/application/saga/ShipOrderStep.ts
import {
  SagaStep,
  StepResult,
  CompensationResult,
  SagaError,
  SagaContext,
  ShippingService
} from '../../domain';

interface ShipOrderResult {
  trackingId: string;
}

/**
 * Saga step for shipping order
 */
export class ShipOrderStep implements SagaStep<ShipOrderResult> {
  readonly name = 'ShipOrder';

  constructor(private readonly shippingService: ShippingService) {}

  async execute(context: SagaContext): Promise<StepResult<ShipOrderResult>> {
    try {
      if (!context.order) {
        return {
          success: false,
          error: new SagaError({
            message: 'Order not found in context',
            code: 'ORDER_NOT_FOUND',
            stepName: this.name,
            isRetryable: false
          })
        };
      }

      const trackingId = await this.shippingService.createShipment(
        context.orderId,
        context.order.customerId
      );

      return {
        success: true,
        data: { trackingId }
      };
    } catch (error) {
      return {
        success: false,
        error: new SagaError({
          message: `Failed to ship order: ${(error as Error).message}`,
          code: 'SHIPPING_FAILED',
          stepName: this.name,
          isRetryable: true
        })
      };
    }
  }

  async compensate(context: SagaContext): Promise<CompensationResult> {
    try {
      if (context.shipmentTrackingId) {
        await this.shippingService.cancelShipment(context.shipmentTrackingId);
      }
      return { success: true };
    } catch (error) {
      return {
        success: false,
        error: new SagaError({
          message: `Failed to cancel shipment: ${(error as Error).message}`,
          code: 'SHIPMENT_CANCELLATION_FAILED',
          stepName: this.name,
          isRetryable: true
        })
      };
    }
  }
}

File: benchmarks/v3/scenarios/05-java-saga/without-mcp/src/main/java/com/example/saga/application/saga/step/ShipOrderStep.java
package com.example.saga.application.saga.step;

import com.example.saga.application.port.ShippingService;
import com.example.saga.application.saga.CompensationResult;
import com.example.saga.application.saga.SagaContext;
import com.example.saga.application.saga.SagaStep;
import com.example.saga.application.saga.StepResult;
import com.example.saga.domain.entity.Order;
import com.example.saga.domain.exception.SagaException;

/**
 * Saga step for shipping the order.
 */
public class ShipOrderStep implements SagaStep<String> {

    private static final String STEP_NAME = "ShipOrder";

    private final ShippingService shippingService;

    public ShipOrderStep(ShippingService shippingService) {
        this.shippingService = shippingService;
    }

    @Override
    public String getName() {
        return STEP_NAME;
    }

    @Override
    public StepResult<String> execute(SagaContext context) {
        try {
            Order order = context.getOrder()
                    .orElseThrow(() -> new SagaException(
                            "Order not found in context",
                            "ORDER_NOT_FOUND",
                            STEP_NAME,
                            false
                    ));

            String trackingId = shippingService.createShipment(
                    context.getOrderId(),
                    order.getCustomerId()
            );

            return StepResult.success(trackingId);
        } catch (SagaException e) {
            return StepResult.failure(e);
        } catch (Exception e) {
            return StepResult.failure(new SagaException(
                    "Failed to ship order: " + e.getMessage(),
                    "SHIPPING_FAILED",
                    STEP_NAME,
                    true,
                    e
            ));
        }
    }

    @Override
    public CompensationResult compensate(SagaContext context) {
        try {
            context.getShipmentTrackingId().ifPresent(
                    shippingService::cancelShipment
            );
            return CompensationResult.success();
        } catch (Exception e) {
            return CompensationResult.failure(new SagaException(
                    "Failed to cancel shipment: " + e.getMessage(),
                    "SHIPMENT_CANCELLATION_FAILED",
                    STEP_NAME,
                    true,
                    e
            ));
        }
    }
}
